internal/chain/engine.go

```diff
@@ -438,127 +389,113 @@ func applySlashes(
-				"[VerifySeal] Not enough voting power in commitSignature from Block Header",
-			)
-		}
-	} else {
-		quorumCount, err := QuorumForBlock(chain, header, reCalculate)
-		if err != nil {
-			return errors.Wrapf(err,
-				"cannot calculate quorum for block %s", header.Number())
-		}
-		if count := utils.CountOneBits(mask.Bitmap); count < int64(quorumCount) {
-			return errors.New(
-				"[VerifyHeaderWithSignature] Not enough signature in commitSignature from Block Header",
-			)
-		}
+	// Verify signature, mask against quorum.Verifier and publicKeys
+	if !qrVerifier.IsQuorumAchievedByMask(mask) {
+		return errors.New("not enough signature collected")
 	}
 	commitPayload := signature.ConstructCommitPayload(chain,
 		header.Epoch(), header.Hash(), header.Number().Uint64(), header.ViewID().Uint64())
 
 	if !aggSig.VerifyHash(mask.AggregatePublic, commitPayload) {
-		return errors.New("[VerifySeal] Unable to verify aggregated signature for block")
+		return errors.New("Unable to verify aggregated signature for block")
 	}
 	return nil
 }
 
-// GetPublicKeys finds the public keys of the committee that signed the block header
-func GetPublicKeys(
-	chain engine.ChainReader, header *block.Header, reCalculate bool,
-) ([]harmony_bls.PublicKeyWrapper, error) {
-	if header == nil {
-		return nil, errors.New("nil header provided")
-	}
-	shardState := new(shard.State)
-	var err error
-	if reCalculate {
-		shardState, _ = committee.WithStakingEnabled.Compute(header.Epoch(), chain)
+func (e *engineImpl) getShardState(chain engine.ChainReader, epoch *big.Int, targetShardID uint32) (*shard.State, error) {
+	// (TODO) For now, when doing cross shard, we need recalcualte the shard state since we don't have
+	//    hard state of other shards
+	if e.needRecalculateStateShard(chain, epoch, targetShardID) {
+		shardState, err := committee.WithStakingEnabled.Compute(epoch, chain)
+		if err != nil {
+			return nil, errors.Wrapf(err, "compute shard state for epoch %v", epoch)
+		}
+		return shardState, nil
+
 	} else {
-		shardState, err = chain.ReadShardState(header.Epoch())
+		shardState, err := chain.ReadShardState(epoch)
 		if err != nil {
-			return nil, errors.Wrapf(
-				err, "failed to read shard state of epoch %d", header.Epoch().Uint64(),
-			)
+			return nil, errors.Wrapf(err, "read shard state for epoch %v", epoch)
 		}
+		return shardState, nil
 	}
+}
 
-	subCommittee, err := shardState.FindCommitteeByID(header.ShardID())
-	if err != nil {
-		return nil, errors.Wrapf(
-			err,
-			"cannot find shard in the shard state at block %d shard %d",
-			header.Number(),
-			header.ShardID(),
-		)
+// only recalculate for non-staking epoch and targetShardID is not the same
+// as engine
+func (e *engineImpl) needRecalculateStateShard(chain engine.ChainReader, epoch *big.Int, targetShardID uint32) bool {
+	if chain.Config().IsStaking(epoch) {
+		return false
+	}
+	return targetShardID != e.shardID
+}
+
+// Support 512 at most validator nodes
+const bitmapKeyBytes = 64
```

Comment:
at most 512 validators per shard?  Can it be changed later?

Reply:
This is simply a sanity field for de-duplication in memory. And sure, it can be changed anytime when needed.